framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,650 @@
+
+/**
+ * SceneToStubWriter provides two static methods named {@code write} that write a {@link AScene} in
+ * stub file format, to a {@link Writer} {@link #write(ASceneWrapper, Writer)}, or to a file {@link
+ * #write(ASceneWrapper, Writer)}. This class is the equivalent of {@code IndexFileWriter} from the
+ * Annotation File Utilities, but outputs the results in the stub file format instead of jaif
+ * format.
+ *
+ * <p>You can use this writer instead of {@code IndexFileWriter} by passing the {@code
+ * -Ainfer=stubs} command-line argument.
+ */
+public final class SceneToStubWriter {
+
+    /** A pattern matching one or more digits. */
+    private static final Pattern digitPattern = Pattern.compile(".*\\$\\d+(\\$.*|$)");
+
+    /** How far to indent when writing members of a stub file. */
+    private static final String INDENT = "    ";
+
+    /**
+     * Writes the annotations in {@code scene} to {@code out} in stub file format.
+     *
+     * @param scene the scene to write out
+     * @param out the Writer to output to
+     */
+    public static void write(ASceneWrapper scene, Writer out) {
+        writeImpl(scene, new PrintWriter(out));
+    }
+
+    /**
+     * Writes the annotations in {@code scene} to the file {@code filename} in stub file format.
+     *
+     * @param scene the scene to write out
+     * @param filename the path of the file to write
+     * @throws IOException if there is trouble writing the file
+     * @see #write(ASceneWrapper, Writer)
+     */
+    public static void write(ASceneWrapper scene, String filename) throws IOException {
+        write(scene, new FileWriter(filename));
+    }
+
+    /**
+     * Returns the part of a binary name that specifies the package.
+     *
+     * @param className the binary name of a class
+     * @return the part of the name referring to the package
+     */
+    // Substrings of binary names are also binary names; the empty string
+    // is a dot-separated identifier (the default package).
+    @SuppressWarnings("signature:return.type.incompatible")
+    private static @DotSeparatedIdentifiers String packagePart(@BinaryName String className) {
+        int lastdot = className.lastIndexOf('.');
+        return (lastdot == -1) ? "" : className.substring(0, lastdot);
+    }
+
+    /**
+     * Returns the part of a binary name that specifies the basename of the class.
+     *
+     * @param className a binary name
+     * @return the part of the name representing the class's name without its package
+     */
+    // Substrings of binary names are also binary names.
+    @SuppressWarnings("signature:return.type.incompatible")
+    private static @BinaryName String basenamePart(@BinaryName String className) {
+        int lastdot = className.lastIndexOf('.');
+        return (lastdot == -1) ? className : className.substring(lastdot + 1);
+    }
+
+    /**
+     * Formats a literal argument of an annotation. Copied from {@code IndexFileWriter#printValue}
```

Comment:
Please don't do this.  Do a minor refactoring in that class rather than doing cut-and-paste copying.
You can create a branch with the same name as this branch, so that the changes to different projects will be tested together.